### ISSUE TITLE
Problem verifying puppet release files

### ISSUE THREAD

**Issue by cyr** [rpgp/rpgp#424]
I've been using this in an application to verify PGP signatures in debian repositories, which has been working well for the most part. But for some reason it does not verify Puppets repository release files, and I can't exactly tell why.

public key:
https://apt.puppetlabs.com/DEB-GPG-KEY-puppet-20250406

inline-signed cleartext message
https://apt.puppetlabs.com/dists/bookworm/InRelease

The id and fingerprints matches, but the verification fails here:

```rust
ensure_eq!(
    &self.signed_hash_value,
    &hash[0..2],
    "signature: invalid signed hash value"
);
```

Any idea why this would be? The gpg utility gives the signature an all clear.

**Comment by hko-s** [comment]
Hello, and thanks for getting in touch!

I wonder if this might be a case where the "2-octet field holding the left 16 bits of the signed hash value" field ([see](https://www.rfc-editor.org/rfc/rfc9580.html#name-versions-4-and-6-signature-)) contains incorrect data, but the cryptographic signature is otherwise correct.

If this is the case, then commenting out the `ensure_eq` in rPGP (the one that you quote) should make the signature validation succeed.

(Assuming this fixes validation, we'd need to figure out if and how to adjust rPGP to allow callers to ignore the 2 bytes of hash digest in the signature packet.)

**Comment by cyr** [comment]
I just tried commenting out that part and let it keep going, but it still stumbles and fails while verifying.

**Comment by hko-s** [comment]
That is intriguing and somewhat troubling :sweat_smile: 

How does the failure look without the ensure?

**Comment by cyr** [comment]
It SEEMS to reach this part:

`rust
crypto::rsa::verify(n.as_bytes(), e.as_bytes(), hash, hashed, sig[0].as_bytes())
`

and fails the verifications there.

**Comment by hko-s** [comment]
Right. That sounds like rPGP's CSF code might be buggy, for this case.
Thanks again for the report, we'll need to dig deeper into this. I will try to do that, but can't promise super quick results.

Either way, this should be resolvable, and we should totally fix it.

Note (mostly) to self:

`cat InRelease | rsop inline-verify DEB-GPG-KEY-puppet-20250406`

reproduces the issue nicely.

**Comment by wiktor-k** [comment]
Yup, I'm getting `error: pgp::errors::Error - "failed reading: cleartext body Error(Error { input: [79, 114, 105,` with this sample code:

```rust
use pgp::composed::cleartext::CleartextSignedMessage;
use pgp::composed::SignedPublicKey;
use pgp::Deserializable as _;
fn main() -> testresult::TestResult {
    let key = SignedPublicKey::from_armor_single(std::fs::File::open(
        "/home/wiktor/tmp/armored-keys.asc",
    )?)?
    .0;
    let msg =
        CleartextSignedMessage::from_armor(std::fs::File::open("/home/wiktor/tmp/InRelease")?)?;
    msg.0.verify(&key)?;

    Ok(())
}
```

**Comment by hko-s** [comment]
> Appreciate it! And I will let you know know if I find anything more on this.

Thanks, great!

CSF is super fiddly about normalization (including, but not limited to line ending encoding). Apparently some edge case is wrong in rPGP, and this message triggers the bug.

My next step will be to add a unit test with these two files to rPGP (followed by a period of cursing, presumably :D)

**Comment by hko-s** [comment]
Ok, I debugged this and implemented a fix in #427.

The issue was that we failed to strip space/tab characters at line endings when calculating the hash digest of the message (see https://www.rfc-editor.org/rfc/rfc9580.html#section-7.1-4).

@cyr: Thanks again for debugging this and the super helpful report!

**Comment by wiktor-k** [comment]
@cyr is your thing open-source? If so could you give a link? I'm always curious what people are using (r)PGP for :)

**Comment by cyr** [comment]
It is!

https://github.com/cyr/aptmirs

It's a debian repository mirroring tool. Written because the other ones (apt-mirror and debmirror in particular) have some problems mirroring repositories that are hosted under specific paths, and not really being up to date on some particulars. And while rsync is nice and good, sometimes your the repository you want to mirror isn't available with rsync. And you might also not care about mirroring every arch!

**Comment by wiktor-k** [comment]
Yep, looks very cool! Thanks for sharing, I'm taking a look :eyes:

**Comment by cyr** [comment]
Enjoy! 😅 I will say that the PGP verification is somewhat new and subject to change. I've realized that having to specify public key by repository gets a little wordy, so I will probably add a general pub key path to pick up any keys found and apply them by key id/fingerprint.

**Comment by hko-s** [comment]
Cool, thanks for the link!

As a (not reasonably actionable, right now) observation:

Verifying signatures in PGP involves quite a number subtleties, which rPGP doesn't offer helpful abstractions to handle.
I'm working on a [higher level library](https://crates.io/crates/rpgpie) on top of rPGP that I hope will conveniently handle these concerns for applications like yours - but I don't think it's quite ready for use, yet (although, it might already cover what you need, if you're feeling adventurous and are not deterred by the prospect of occasional breaking API changes).

Specifically, it's relatively fiddly to make sure if the primary or the subkeys are actually valid, and are ok to use for validation of data signatures (e.g.: if subkeys are validly bound to the primary; that any given key is neither expired nor revoked; that the key has the "data signature" feature flag, which signals that it is "allowed" to issue data signatures, that no ancient broken cryptographic primitives are used, ...).

**Comment by cyr** [comment]
That sounds very interesting! I did note that those subtleties were there, but I chose to not tinker with those details with my initial implementation, but I'd be very interested in relaying those concerns to a higher level library. I'll take a look!

**Comment by hko-s** [comment]
> That sounds very interesting! I did note that those subtleties were there, but I chose to not tinker with those details with my initial implementation, but I'd be very interested in relaying those concerns to a higher level library. I'll take a look!

Cool!

This is a place where I'm using rpgpie to validate CSF signatures:

https://codeberg.org/heiko/rsop/src/commit/871bc394d7288af699b43bf042990fbad250be35/rpgpie-sop/src/cmd/inline_verify.rs#L125-L128

It's relatively messy code, but maybe still helpful as an outline of the general approach.

I'm more than happy to hear from you if you decide to dig into rpgpie, and happy to help - including if you're missing functionality. Fleshing out that higher level library will be a priority of mine over the coming months.

**Comment by cyr** [comment]
I'll definitely read through it. And perhaps adapt some of my own validation in the meantime. 😅

As a general use case (though it might already be obvious), I would ideally want an api that takes an impl Read for both public keys and signatures/messages/inlined-signatures and returning a vaguely accurate error if validation fails. I did note that you do not support streaming reads right now, which is fine for my case at the moment admittedly - the release files aren't overly large typically - but I do try to avoid it whenever possible. I'm assuming this is because of the rpgp api as it stands right now? Either way, not a big deal, just mentioning what an ideal api would look like in my mind 😌

**Comment by hko-s** [comment]
As a tiny bit more context (apologies for info-dumping):

I started writing rpgpie out of curiosity, to plug rPGP into the "OpenPGP interop test suite": https://sequoia-pgp.gitlab.io/openpgp-interoperability-test-suite/results.html

Over the course of a few weeks of hacking on that, I concluded that rPGP (which I didn't yet know much about, at the time) performs extremely well, and is an easy codebase to learn. I found - and fixed - a few bugs, and my experience of getting them upstreamed into rPGP was great.

Now that the interop tests are looking good, and now that rPGP implements the new RFC 9580, my next priority is to evolve rpgpie from a hacky piece of glue code into a nice, and hopefully general purpose, API for a broad range of OpenPGP applications.

There's a lot to do still, but in terms of its functionality, and the checks it performs, rpgpie is already covering quite a lot of ground (as demonstrated by the results in the interop suite).
